### PR TITLE
esn#11 Add missing dependencies for tests

### LINKED ISSUE #11: Developer Experience: setup tests
### Step by step on how to set up tests for a module: 

1. Create a new webpack config file like [this one](https://github.com/OpenPaaS-Suite/esn-frontend-calendar/blob/d02ba911ed6d9be7194b10ec7a24c93845ff2b7c/webpack.test.js)

2. Copy old test folder from old repository, remove all unnecessary files and leave only frontend test related files there.

3.  Update the [karma.conf.js file](https://github.com/OpenPaaS-Suite/esn-frontend-calendar/blob/c7a23b7c6c6bf4c82b55e59b3c27eb2393eb2032/test/config/karma.conf.js)
 - `config.files` will have a file named index.test.js at top. This file will play the role as our webpack entry point file.
 - Note that we now use FirefoxHeadless instead of PhantomJS.

4. Update package.json:
 - Add this script: `"test": "./node_modules/karma/bin/karma start test/config/karma.conf.js"`

5. Create index.test.js file inside `src` folder. This file will require all necessary dependencies, mdule files & test files. See [here](https://github.com/OpenPaaS-Suite/esn-frontend-calendar/blob/c7a23b7c6c6bf4c82b55e59b3c27eb2393eb2032/src/index.test.js)

6. Try to run the test and fix the problems ( if any )

### Some common problems appear when working on this issue (Updating...)

1. Missing chai / sinon / ...
For those libraries please install them via npm and require them in `index.test.js`

2. `jadeTemplates`
Remove all `jadeTemplates` module since we have loaded all of our pug files via webpack.

3. Unexpected request
 Most likely those unexpected requests should be /api/i18n or /api/jwt/generate. In this case, some `run` functions inside `esn-frontend-common-libs` have caused this. To resolve this, first detect where they have been called.
Eg: For `/api/i18n`, it was called in a file named `i18n-loader.service.js` inside `i18n` module.
Then disable the run function by exclude the file ( If you can ) or, in this case, replace the file with a new mock file. I created a mock file called `i18n-loader.js` inside `mocks` folder in `test`. After that use `NormalModuleReplacementPlugin` of webpack to replace the original file with the mock one.

4. `$q is undefined`
In some test cases, we have some `beforeEach` hooks that are using $q to mock sth. Eg:

var abcMock = {
   get: $q.when({})
};

When webpack wraps things & comes to this point, it will throw an error: `$q is undefined`. `$q` was originally a built-in service of AngularJS therefore we cannot use it without injecting it.

Back to the old days (old code base), we have a file called `injector.js` inside `mocks` folder in `test`. What this file does is: 
 - Declare an AngularJS module `esn.test.injector` & a `run` function. This function will inject dependencies such as `$q` and assign it to global variable.
 - Create a global `beforeEach` hook, which will inject module `esn.test.injector` in every test case.
So theoretically, we will have our `$q` global & no more `$q is undefined`.

However, I discovered that this won't work as expected. The lifecycle of mocha testing should look like this:

`beforeEach global hook`
       `beforeEach for test file`
              `beforeEach in describe block`
                        `test case (it)`

So the `run` function of `esn.test.injector` won't run right after the beforeEach global hook executed, but instead run when it comes to test case. Therefore in `beforeEach` hook, `$q` is still undefined.

To make things work, I came up with a solution. I require & use `$q` from `node_modules/q/q.js` as global $q.

var $q = require('q/q.js');
window.$q = $q;

Please let me know if you have any better approach.

5. Unknown provider
Most likely it is because we have missed to inject some modules to our tests. Eg: `esn.calendar.libs` is a new module that was created to store services & common functions. To use them you need to inject `esn.calendar.libs` to our tests:
`angular.mock.module('esn.calendar.libs')`

### PR DESCRIPTION
**Pull request by ppppbn**
Partially resolve https://github.com/OpenPaaS-Suite/esn/issues/11

### CODE REVIEW

File: src/frontend/components/chai-datetime/.bower.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "chai-datetime",
```
Comment (tuanlc) [inline src/frontend/components/chai-datetime/.bower.json:2]: You can install `chai-datetime` by npm, so no need to copy it's code in component

File: src/frontend/components/chai/.bower.json [hunk 2]
```diff
@@ -0,0 +1,37 @@
+{
+  "name": "chai",
```
Comment (tuanlc) [inline src/frontend/components/chai/.bower.json:2]: Same for `chai`, you can install it by `npm`

File: src/frontend/components/sinon-chai/.bower.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "sinon-chai",
```
Comment (tuanlc) [inline src/frontend/components/sinon-chai/.bower.json:2]: same here `sinon-chai` can be installed by `npm`

File: src/frontend/components/angular-mocks/.bower.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "angular-mocks",
```
Comment (tuanlc) [inline src/frontend/components/angular-mocks/.bower.json:2]: You can install this by `npm`, so no need to copy its code

**Review by tuanlc — COMMENTED** [review]
For the pakages that you can install them by `npm`, no need to copy their code bases

File: src/frontend/components/angular-mocks/bower.json [hunk 5]
```diff
@@ -0,0 +1,9 @@
+{
+  "name": "angular-mocks",
+  "version": "1.3.20",
```
Comment (itswillta) [inline src/frontend/components/angular-mocks/bower.json:3]: It seems that `angular-mocks` can be installed via npm as well: https://www.npmjs.com/package/angular-mocks

**Comment by ppppbn** [comment]
Close this pull request because all the packages / dependencies needed can and were decided to be installed via npm.

**Pull request closed without merging**